File: frontend/src/services/monitor.ts
import { apiGet, apiPost, apiPut, apiDelete } from './api';

export interface Metric {
    id: string;
    server_id?: string;
    metric_name: string;
    value: number;
    unit?: string;
    labels?: Record<string, string>;
    collected_at: string;
}

export interface MetricPoint {
    bucket: string;
    avg: number;
    min: number;
    max: number;
    count: number;
}

export interface AlertRule {
    id: string;
    name: string;
    metric_name: string;
    condition: string; // gt / lt / gte / lte
    threshold: number;
    duration: number;
    severity: string; // info / warning / critical
    server_id?: string | null;
    is_enabled: boolean;
    cooldown: number;
    created_at: string;
    updated_at: string;
}

export interface AlertEvent {
    id: string;
    rule_id?: string;
    rule_name?: string;
    server_id?: string;
    metric_name?: string;
    value: number;
    threshold: number;
    condition?: string;
    severity: string;
    message?: string;
    status: string; // firing / acknowledged
    triggered_at: string;
    acknowledged_by?: string;
    acknowledged_at?: string;
}

export const monitorApi = {
    queryMetrics: (params: { server_id?: string; name?: string; since?: string; limit?: number; aggregate?: string }) =>
        apiGet<{ code: number; data: { items: (Metric | MetricPoint)[]; total?: number } }>(
            `/monitor/metrics${toQuery(params)}`
        ).then(r => r.data),
    latestMetrics: (serverId?: string) =>
        apiGet<{ code: number; data: { items: Metric[]; total: number } }>(
            `/monitor/metrics/latest${serverId ? `?server_id=${encodeURIComponent(serverId)}` : ''}`
        ).then(r => r.data),
    ingestMetrics: (metrics: Partial<Metric>[]) =>
        apiPost<{ code: number; data: any }>('/monitor/metrics/ingest', { metrics }).then(r => r.data),

    listRules: () =>
        apiGet<{ code: number; data: { items: AlertRule[]; total: number } }>('/monitor/alert-rules').then(r => r.data),
    createRule: (rule: Partial<AlertRule>) =>
        apiPost<{ code: number; data: AlertRule }>('/monitor/alert-rules', rule).then(r => r.data),
    updateRule: (id: string, rule: Partial<AlertRule>) =>
        apiPut<{ code: number; data: AlertRule }>(`/monitor/alert-rules/${id}`, rule).then(r => r.data),
    deleteRule: (id: string) => apiDelete(`/monitor/alert-rules/${id}`),

    listAlerts: (status?: string) =>
        apiGet<{ code: number; data: { items: AlertEvent[]; total: number } }>(
            `/monitor/alerts${status ? `?status=${status}` : ''}`
        ).then(r => r.data),
    ackAlert: (id: string) => apiPost(`/monitor/alerts/${id}/ack`, {}),
};

function toQuery(params: Record<string, any>): string {
    const qs = Object.entries(params)
        .filter(([, v]) => v !== undefined && v !== null && v !== '')
        .map(([k, v]) => `${k}=${encodeURIComponent(v)}`)
        .join('&');
    return qs ? `?${qs}` : '';
}

File: backend/internal/models/monitor.go
package models

import "time"

// Metric is a single collected monitoring sample.
type Metric struct {
	ID          string            `json:"id"`
	ServerID    string            `json:"server_id,omitempty"`
	MetricName  string            `json:"metric_name"`
	Value       float64           `json:"value"`
	Unit        string            `json:"unit,omitempty"`
	Labels      map[string]string `json:"labels,omitempty"`
	CollectedAt time.Time         `json:"collected_at"`
}

// AlertRule defines a threshold condition on a metric.
type AlertRule struct {
	ID        string    `json:"id"`
	Name      string    `json:"name"`
	MetricName string   `json:"metric_name"`
	Condition string    `json:"condition"` // gt, lt, gte, lte
	Threshold float64   `json:"threshold"`
	Duration  int       `json:"duration"` // seconds; 0 = immediate
	Severity  string    `json:"severity"` // info, warning, critical
	ServerID  *string   `json:"server_id,omitempty"`
	IsEnabled bool      `json:"is_enabled"`
	// IsEnabledSet distinguishes "explicitly false" from "omitted" (JSON
	// cannot express both on a plain bool). When omitted the rule is enabled.
	IsEnabledSet bool `json:"-"`
	Cooldown  int       `json:"cooldown"` // seconds between repeat alerts
	// NotificationChannels is a JSON array of notification_channel ids bound to
	// the rule (Phase 7). Empty → dispatcher selects channels by template.
	NotificationChannels string `json:"notification_channels,omitempty"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// AlertEvent is a fired alert instance.
type AlertEvent struct {
	ID             string     `json:"id"`
	RuleID         string     `json:"rule_id,omitempty"`
	RuleName       string     `json:"rule_name,omitempty"`
	ServerID       string     `json:"server_id,omitempty"`
	MetricName     string     `json:"metric_name,omitempty"`
	Value          float64    `json:"value"`
	Threshold      float64    `json:"threshold"`
	Condition      string     `json:"condition,omitempty"`
	Severity       string     `json:"severity"`
	Message        string     `json:"message,omitempty"`
	Status         string     `json:"status"` // firing, acknowledged, resolved
	TriggeredAt    time.Time  `json:"triggered_at"`
	AcknowledgedBy string     `json:"acknowledged_by,omitempty"`
	AcknowledgedAt *time.Time `json:"acknowledged_at,omitempty"`
}

// MetricPoint is an aggregated datapoint for charts.
type MetricPoint struct {
	Bucket string  `json:"bucket"` // ISO timestamp of bucket start
	Avg    float64 `json:"avg"`
	Min    float64 `json:"min"`
	Max    float64 `json:"max"`
	Count  int     `json:"count"`
}
